runtime/src/cost_model.rs

```diff
@@ -166,18 +171,26 @@ impl CostModel {
             // to keep the same behavior, look for builtin first
             if let Some(builtin_cost) = BUILT_IN_INSTRUCTION_COSTS.get(program_id) {
                 builtin_costs = builtin_costs.saturating_add(*builtin_cost);
-            } else {
-                let instruction_cost = self.find_instruction_cost(program_id);
-                trace!(
-                    "instruction {:?} has cost of {}",
-                    instruction,
-                    instruction_cost
-                );
-                bpf_costs = bpf_costs.saturating_add(instruction_cost);
+            } else if !compute_budget::check_id(program_id) {
+                bpf_costs = bpf_costs.saturating_add(DEFAULT_INSTRUCTION_COMPUTE_UNIT_LIMIT.into());
             }
             data_bytes_len_total =
                 data_bytes_len_total.saturating_add(instruction.data.len() as u64);
         }
+
+        // calculate bpf cost based on compute budget instructions
+        let mut budget = ComputeBudget::default();
+        let result = budget.process_instructions(
+            transaction.message().program_instructions_iter(),
+            true,
+            false, /*not support request_units_deprecated*/
+        );
+
+        // if tx contained user-space instructions and a more accurate estimate available correct it
+        if bpf_costs > 0 && result.is_ok() {
+            bpf_costs = budget.compute_unit_limit
```

Comment:
This is subtle that user requested cu suppose to cover not just BPF execution cost, but everything else (signature, write lock, etc), it is total CUs for entire transaction. So maybe changes should be in `calculate_cost()` function to set tx_cost = get_requested_compute_units(sanitized_transaction). But we still want to keep write_account map, account data size logic and vote limit etc.

Reply:
could you clarify? i thought the difference is:
cost units = gross, superset of both, includes account write & signature, ledger data, 48M limit per block
compute units = net, metric strictly of program execution cost

the compute budget instruction is called set_compute_unit_limit, it should only limit the compute units (net) aka bpf costs.
fee model uses priority * compute units (net)